### PR TITLE
More IPython like behaviour

### PR DESCRIPTION
**Pull request by fmagin**
This PR will consist of multiple commits for separate concerns:

One thing I am missing currently is working tab completion for the attributes of the object that is being bridged. The first commit of this PR simply overwrites the __dir__ method of a BridgedObject and adds all the `_bridge_attrs` which seems to work as I want. Tab completion now works in the IPython shell, the only issue being Jedi which I think can be ignored for now [0]

[0] When jedi is activated tab complete only seems to work if at least one letter is typed e.g. `currentProgram.u<tab>` instead of just `currentProgram.<tab>`

### CODE REVIEW

**Comment by fmagin** [comment]
93d43fd is a bit opinionated  but I think it is more sensible that the first thing in the BridgedObject representation should be the representation of the bridged object. The type is now the second entry. This also changes the method that the class e.g. `BridgedObject` isn't hardcoded anymore but retrieved during the call, which allows the objects inheriting to simply keep this method. 

Example:
```python
In [5]: currentProgram                                                                            
Out[5]: <BridgedObject('morph - .ProgramDB', type=ProgramDB, handle=2983110f-888b-4d5f-b511-44ae1b38fb51)>
```

**Comment by fmagin** [comment]
The above fails with BridgedCallable for reasons beyond me. Because of that `BridgedCallable` still has the old __repr__ method. The message `TypeError: __repr__ requires at least one argument` is fairly confusing because to my understanding __repr__ should not need any arguments

```
---------------------------------------------------------------------------
BridgeException                           Traceback (most recent call last)
<ipython-input-1-f05db18a4cde> in <module>
      1 import ghidra_bridge
----> 2 b = ghidra_bridge.GhidraBridge(namespace=globals()) # creates the bridge and loads the flat API into the global namespace
      3 print(getState().getCurrentAddress().getOffset())
      4 # ghidra module implicitly loaded at the same time as the flat API
      5 ghidra.program.model.data.DataUtilities.isUndefinedData(currentProgram, currentAddress)

~/Projects/ghidra_bridge/ghidra_bridge/ghidra_bridge.py in __init__(self, connect_to_host, connect_to_port, loglevel, namespace, interactive_mode, response_timeout)
     55             # track the namespace we loaded with - if we're part of an __enter__/__exit__ setup, we'll use it to automatically unload the flat api
     56             self.namespace = namespace
---> 57             self.get_flat_api(namespace=self.namespace)
     58 
     59     def get_flat_api(self, namespace=None):

~/Projects/ghidra_bridge/ghidra_bridge/ghidra_bridge.py in get_flat_api(self, namespace)
     80                 # first, manually update all the current* values (this allows us to get the latest values, instead of what they were when the server started
     81                 tool = remote_main.state.getTool()  # note: tool shouldn't change
---> 82                 listing_panel = get_listing_panel(tool, remote_main.ghidra)
     83                 locn = listing_panel.getProgramLocation()
     84                 # set the values as overrides in the bridged object - this prevents them from being changed in the remote object

~/Projects/ghidra_bridge/ghidra_bridge/ghidra_bridge.py in get_listing_panel(tool, ghidra)
     17 def get_listing_panel(tool, ghidra):
     18     """ Get the code listing UI element, so we can get up-to-date location/highlight/selection """
---> 19     cvs = tool.getService(ghidra.app.services.CodeViewerService)
     20     return cvs.getListingPanel()
     21 

~/Projects/ghidra_bridge/ghidra_bridge/bridge.py in __call__(self, *args, **kwargs)
   1072 
   1073     def __call__(self, *args, **kwargs):
-> 1074         return self._bridge_conn.remote_call(self._bridge_handle, *args, **kwargs)
   1075 
   1076     def __get__(self, instance, owner):

~/Projects/ghidra_bridge/ghidra_bridge/bridge.py in remote_call(self, handle, *args, **kwargs)
    640     def remote_call(self, handle, *args, **kwargs):
    641         self.logger.debug(
--> 642             "remote_call: {}({},{})".format(handle, args, kwargs))
    643 
    644         serial_args = self.serialize_to_dict(args)

~/Projects/ghidra_bridge/ghidra_bridge/bridge.py in __repr__(self)
   1031 
   1032     def __repr__(self):
-> 1033         return self._bridged_get("__repr__")()
   1034 
   1035     def __dir__(self):

~/Projects/ghidra_bridge/ghidra_bridge/bridge.py in __call__(self, *args, **kwargs)
   1072 
   1073     def __call__(self, *args, **kwargs):
-> 1074         return self._bridge_conn.remote_call(self._bridge_handle, *args, **kwargs)
   1075 
   1076     def __get__(self, instance, owner):

~/Projects/ghidra_bridge/ghidra_bridge/bridge.py in remote_call(self, handle, *args, **kwargs)
    647                                           ARGS: serial_args, KWARGS: serial_kwargs}}
    648 
--> 649         return self.deserialize_from_dict(self.send_cmd(command_dict))
    650 
    651     def local_call(self, args_dict):

~/Projects/ghidra_bridge/ghidra_bridge/bridge.py in deserialize_from_dict(self, serial_dict)
    529         elif serial_dict[TYPE] == EXCEPTION:
    530             raise BridgeException(self.deserialize_from_dict(serial_dict[MESSAGE]), BridgedObject(
--> 531                 self, serial_dict[VALUE]))
    532         elif serial_dict[TYPE] == BRIDGED:
    533             return self.get_object_by_handle(serial_dict[VALUE])

BridgeException: ('__repr__ requires at least one argument', TypeError('__repr__ requires at least one argument',))
```

File: ghidra_bridge/bridge.py [hunk 1]
```diff
@@ -1084,9 +1084,6 @@ class BridgedIterable(BridgedObject):
     def __iter__(self):
         return self._bridged_get("__iter__")()
 
-    def __repr__(self):
```
Comment (justfoxing) [inline ghidra_bridge/bridge.py]: Nice, much cleaner this way

**Comment by justfoxing** [comment]
Ahh, I see what's going on with BridgedCallable. It's actually specifically when the ghidra.app.services.CodeViewerService class is being repr'd as an argument in the remote_call logging format string on 642. I think most BridgedCallables actually work fine, but I have seen issues with javapackage and class objects with \_\_str__ before (I think it a class vs instance object thing).

What we actually need to do is get the remote type, and call repr against that, with the object as the argument (e.g., so type(ghidra_class).__repr__(ghidra_class) - probably implemented in the bridge as something like `self._bridged_type().__str__(self)`). It'll need a remote_type() command in the bridge to pass back the type object. I can probably knock that out pretty quick.

**Comment by justfoxing** [comment]
Also, noting some logging output on the server-side that'd be nice to avoid:


```
b = ghidra_bridge.GhidraBridge(namespace=globals())
currentAddress
```
AttributeError: 'ghidra.program.model.address.GenericAddress' object has no attribute '__dir__'
AttributeError: 'ghidra.program.model.address.GenericAddress' object has no attribute '_ipython_canary_method_should_not_exist_'

```
b = ghidra_bridge.GhidraBridge()
ghidra = b.get_ghidra_api()()
```
ghidra.<tab>
    AttributeError: 'javapackage' object has no attribute '__delattr__'

**Comment by justfoxing** [comment]
Phew. Okay, sorry about the scads of comments and editing/re-editing - took me a little while to get my thoughts clear. I'll stop commenting now.

HMU if anything's unclear.

File: ghidra_bridge/bridge.py [hunk 2]
```diff
@@ -1027,7 +1027,7 @@ def __str__(self):
         return self._bridged_get("__str__")()
 
     def __repr__(self):
-        return "<BridgedObject({}, handle={})>".format(self._bridge_type, self._bridge_handle)
+        return "<{}('{}', type={}, handle={})>".format(type(self).__name__, self._bridged_get("__repr__")(), self._bridge_type, self._bridge_handle)
```
[the remark this answers is not in the document]
Comment (fmagin) [reply ghidra_bridge/bridge.py]: That sounds good and given the latency of bridged calls should be a lot faster too. I'd like some optional debug code somewhere though that asserts that the `_bridge_repr` and the current result of `repr()` on the server haven't diverged. Maybe this should be added for all cases of performance shortcuts that might create such issues. So one could just set some global debug variable and find such issues at the price of performance.

**Comment by fmagin** [comment]
Concerning the warnings in the log: I can't exactly reproduce this (while on 93d43fd2e759829c8e2a94017033efe032a70cea), I only get
```
AttributeError: 'ghidra.program.model.address.GenericAddress' object has no attribute '_ipython_canary_method_should_not_exist_'
```
I am not sure if we can get rid of that, this is a feature of IPython. Maybe we can hide that specific warning/traceback?

Concerning the other issues I will look into them later this evening maybe. Any reason https://github.com/justfoxing/ghidra_bridge/pull/18 isn't merged yet? I would need to rebase my branch on that to continue working

File: ghidra_bridge/bridge.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fmagin) [reply ghidra_bridge/bridge.py]: Addressed in 3c976c4

File: ghidra_bridge/bridge.py [hunk 3]
```diff
@@ -338,7 +339,7 @@ def __init__(self, local_obj):
         self.attrs = dir(local_obj)
 
     def to_dict(self):
-        return {HANDLE: self.handle, TYPE: type(self.local_obj).__name__, ATTRS: self.attrs}
+        return {HANDLE: self.handle, TYPE: type(self.local_obj).__name__, ATTRS: self.attrs, REPR: repr(self.local_obj)}
```
Comment (justfoxing) [inline ghidra_bridge/bridge.py]: 👍 Looks good.

**Comment by justfoxing** [comment]
I'll try to give it a test tonight, but that looks very nice. I reckon you can remove the BridgedCallable repr function now, this version shouldn't run into the issue.

With respect to a debug option - yeah, good idea. I was thinking maybe a feature flag that'll just let you say "always do a bridged repr". Once we get this PR done, I can add that in (I want to go look up feature flag strategies for python first, to make sure I pick something easy to maintain).

**Comment by fmagin** [comment]
There is a __debug__ variable in Python that activates asserts. The issue is that it defaults to True unless -O is passed to the python interpreter, so it would be on for basically everyone using it.

I think something like an assert would be better to explicitly mark that we are making an assumption and thrown an error (or at least a warning) that can be investigated if the cached and the bridged repr diverge.
For example something where the repr changes with every run anyway, so that would never be noticed when restarting the interpreter/bridge.

I will test and remove the repr of the callable later.

**Comment by justfoxing** [comment]
Your new changes tested out okay. Once you push up a commit taking out the callable repr, I'm good to merge this in. I'll tackle the debug option and the warning messages from ipython in separate commits (I think that's related to working out the "tab-complete needing at least one character" issue).

Do you want a callout in a "Contributors" section in the README, or would you prefer to stay low-key?

**Comment by fmagin** [comment]
The "tab-complete needing at least one character" is due to IPython's Jedi completion, but the canary method warning still appears with Jedi deactivated so it is not directly related, but rather both a case of "IPython doing fancy barely documented things"

Callout in the Contributors section would be nice, I have a bunch of other things I am thinking about implementing so I hope that I will have some more contributions beside this.

**Comment by fmagin** [comment]
So, things are getting more interesting:

Turns out we already have a case where `self._bridged_get("__repr__")()` is not the same as `self._bridge_repr`: Probably every callable. The interesting thing is though that what we are caching is what is actually useful so I am implementing it like this nonetheless.



`currentProgram.getFunctionManager._bridg_repr` => `<bound method ghidra.program.database.ProgramDB.getFunctionManager of morph - .ProgramDB>`

`currentProgram.getFunctionManager._bridged_get("__repr__")()` => `<instancemethod object at 0x119b>`

**Comment by justfoxing** [comment]
Cheers for that. It looks like there's some difference between `repr(test_object)` and `test_object.__repr__()` for some Jython objects in this case. 

Despite that, I don't think it's worth holding this up - I agree with you that the representation you've got is more useful as is. I'll make a note and investigate down the track.

I'll merge this PR in now, and follow up with a fix for the Jedi tab autocomplete issue, then cut a new release.

Thanks for the contrib :) When you are thinking of more contributions, hit me up in an issue or email, just so we can make sure it'll fit in and won't run into any issues with changes I'm planning.

**Pull request merged**